urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/yuccasrv/Yucca.java

```diff
@@ -0,0 +1,1099 @@
+     *         the given ID is already bound; the listener set is left unchanged
+     *         then.
+     **/
+    private boolean addConnectionHandlerForSocket( ConnectionHandler handler ) {
+
+	synchronized( this.singleConnectionDispatcher ) {
+	    
+	    ConnectionHandler oldHandler = this.singleConnectionDispatcher.get( handler.getUUID() );
+	    if( oldHandler == null ) {
+
+		// OK, does not yet exist
+		this.singleConnectionDispatcher.put( handler.getUUID(), handler );
+
+		return true;
+	    } else if( handler == oldHandler ) {
+
+		// No change
+		return true;
+
+	    } else {
+
+		this.getLogger().finest( "Failed to set connection handler '"+handler.getUUID()+"' for socket. A different handler for the same socket already exists!" );
+		return false;
+
+	    }
+
+	}
+	
+    }
+
+    /**
+     * As the registration of new listening sockets is non-deterministic (starting new
+     * threads) the binding and retrieving of connection handlers should be 
+     * thread safe. 
+     * Only use this method if you want to access the singleConnectionDispatcher!
+     * (it's synchronized)
+     *
+     * @param handlerID The ID you want to get the handler for (must be unique!).
+     * 
+     * @return the handler itself if the handler was found, null otherwise.
+     **/
+    private ConnectionHandler getConnectionHandlerByUUID( UUID handlerID ) {
+
+	//System.out.println( "ConnectionHandlders="+this.singleConnectionDispatcher );
+
+	synchronized( this.singleConnectionDispatcher ) {
+
+	    return this.singleConnectionDispatcher.get( handlerID ); 
+
+	}
+
+    }
+
+    /**
+     * Get the global logger.
+     *
+     * This is the same logger as returned by LogManager.getLogger( Constants.DEFAULT_LOGGER_NAME );
+     **/
+    public Logger getLogger() {
+	return this.logger;
+    }
+
+    public AbstractCommandLine<Command> getCommandLine() {
+	return this.yuccaLine;
+    }
+
+    public void performQuit( boolean forceQuit ) {
+	// Release all sockets 
+	// !!! THIS METHOD HAS NO IMPLEMENTATION YET !!!
+	this.logger.info( "Closing server sockets." );
+	this.bindManager.closeAllServerSockets();
+	this.bindManager.finalize( 5L, java.util.concurrent.TimeUnit.SECONDS );
+
+	// Remove listeners
+	// ...
+
+	// Force quit all listeners
+	
+
+	// Stop command line
+	// ...
+	if( this.yuccaLine != null ) {
+	    this.logger.info( "Stopping command line." );
+	    this.yuccaLine.interrupt();
+	}
+
+	System.out.println( "Goodbye." );
+
+	if( forceQuit )
+	    System.exit( 0 );
+    }
+
+    public void performStatus() {
+	
+	String statusString = this.bindManager.getStatusString();
+	this.getLogger().log( Level.INFO,
+			      "Socket manager status:\n " + statusString
+			      );
+	
+    }
+
+    public void performPrintLicense() {
+	String licenseFile = "gpl-3.0.txt";
+	BufferedReader reader = null;
+	try {
+
+	    reader = new BufferedReader( new FileReader( new File(licenseFile) ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 239, 241. The resource is closed at line 247. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).